[DevOps] Allow to use vsdrops AND xamarin storage. (#8916)
xamarin/xamarin-macios: tools/devops/device-tests/README.md

      fi
    fi

    make -C builds download .stamp-mono-ios-sdk-destdir -j
    MONO_ENV_OPTIONS=--trace=E:all make -C tests vsts-device-tests
  env:
    WORKING_DIR: $(System.DefaultWorkingDirectory) 
    XAMARIN_STORAGE_PATH: $(XAMARIN_STORAGE_PATH)
    TESTS_EXTRA_ARGUMENTS: ${{ parameters.testsLabels }}
    USE_XAMARIN_STORAGE: '${{ parameters.useXamarinStorage }}'
    VSDROPS_URI: '${{ parameters.vsdropsPrefix }}/$(Build.BuildNumber)/$(Build.BuildId);/' # uri used to create the vsdrops index using full uri
  displayName: 'Run tests'
  name: runTests 
  timeoutInMinutes: 600
```

There are two details we have to pay attention here, first:

```yaml
name: runTests 
```

We need to set the name of the step. This should not be confused with the displayName which is simply a string used in the vsts webpage.
Later we can access the output variables of the step in the following way:

```yaml

XAMARIN_STORAGE_PATH: $[ dependencies.tests.outputs['runTests.XAMARIN_STORAGE_PATH'] ]
```

The other piece of code is

```yaml

echo '##vso[task.setvariable variable=XAMARIN_STORAGE_FAILED;isOutput=true]true'
echo '##vso[task.setvariable variable=XAMARIN_STORAGE_PATH;isOutput=true]""'
```

> As you can see, we use **isOutput=true** which means that the environment variable will be registered as an output variable of the step.

Reviewer: Not sure where it would be best to put, but I'm wondering whether you could add a small note about making sure that vars required by VSDrops are accessible across machines?

- common (under templates/common-variables.yml): Contains a set of group variables that have been created by release engineering and that contain all the different secrets needed by the pipeline.
- lab (under templates/[ddfun|cambridge]-variables: Contains the different configurations per lab. That includes the pools to use and the type of storage.

> There are vertain enviroment variables that are created by steps in the first job that are later shared with the second job. In order to do that you

Reviewer 2: vertain -> certain 
enviroment -> environment

> There are vertain enviroment variables that are created by steps in the first job that are later shared with the second job. In order to do that you
> have to ensrue that you do the following:

Reviewer 3: ```suggestion
There are certain enviroment variables that are created by steps in the first job that are later shared with the second job. In order to do that you
have to ensure that you do the following:
```

> have to ensrue that you do the following:

Reviewer 2: ensrue -> ensure

```yaml
- bash: |
    set -x
    set -e
    echo 'Use xamarin storage $USE_XAMARIN_STORAGE'

    cd $WORKING_DIR/xamarin-macios
    if [[ "$USE_XAMARIN_STORAGE" == "True" ]]; then
      EC=0
      ssh builder@xamarin-storage "mkdir -p /volume1/storage/$XAMARIN_STORAGE_PATH" || EC=$?
      if [ $EC -eq 0 ]; then
        echo '##vso[task.setvariable variable=XAMARIN_STORAGE_FAILED;isOutput=true]false'
        echo '##vso[task.setvariable variable=XAMARIN_STORAGE_PATH;isOutput=true]$XAMARIN_STORAGE_PATH'
        export TESTS_PERIODIC_COMMAND="--periodic-interval 10 --periodic-command rsync --periodic-command-arguments '-avz -e \"ssh\" $PWD/jenkins-results builder@xamarin-storage:/volume1/storage/$XAMARIN_STORAGE_PATH'"
      else
        echo '##vso[task.setvariable variable=XAMARIN_STORAGE_FAILED;isOutput=true]true'
        echo '##vso[task.setvariable variable=XAMARIN_STORAGE_PATH;isOutput=true]""'
      fi
    fi

>     make -C builds download -j || true
>     make -C builds downloads -j || true
>     make -C builds .stamp-mono-ios-sdk-destdir -j || true

Reviewer 3: This can be simplified now that this code is in our own repo (the `downloads` target doesn't exist anymore, but it was still in Azure Devops because the pipeline could be used to build old commits).

```suggestion
    make -C builds download .stamp-mono-ios-sdk-destdir -j
```

    MONO_ENV_OPTIONS=--trace=E:all make -C tests vsts-device-tests
  env:
    WORKING_DIR: $(System.DefaultWorkingDirectory) 
    XAMARIN_STORAGE_PATH: $(XAMARIN_STORAGE_PATH)
    TESTS_EXTRA_ARGUMENTS: ${{ parameters.testsLabels }}
    USE_XAMARIN_STORAGE: '${{ parameters.useXamarinStorage }}'
    VSDROPS_URI: '${{ parameters.vsdropsPrefix }}/$(Build.BuildNumber)/$(Build.BuildId);/' # uri used to create the vsdrops index using full uri
  displayName: 'Run tests'
  name: runTests 
  timeoutInMinutes: 600
```

There are two details we have to pay attention here, first:

```yaml
name: runTests 
```

> We need to set the name of the step. This should not be confused with the dispplayName which is simply a string used in the vsts webpage.

Reviewer 2: dispplayName -> displayName

> We need to set the name of the step. This should not be confused with the dispplayName which is simply a string used in the vsts webpage.

Reviewer 3: ```suggestion
We need to set the name of the step. This should not be confused with the displayName which is simply a string used in the vsts webpage.
```

> We need to set the name of the step. This should not be confused with the dispplayName which is simply a string used in the vsts webpage.

Reviewer 4: ```suggestion
We need to set the name of the step. This should not be confused with the displayName which is simply a string used in the vsts webpage.
```

Later we can access the output variables of the step in the following way:

```yaml

XAMARIN_STORAGE_PATH: $[ dependencies.tests.outputs['runTests.XAMARIN_STORAGE_PATH'] ]
```

The other piece of code is

```yaml

echo '##vso[task.setvariable variable=XAMARIN_STORAGE_FAILED;isOutput=true]true'
echo '##vso[task.setvariable variable=XAMARIN_STORAGE_PATH;isOutput=true]""'
```

> As you can see, we use **isOutput=true** which means that the enviroment variable will be registered as an output variable of the step.

Reviewer 2: enviroment -> environment

## Storage

The html report can be stored in two different places:

* vsdrops
* xamarin-storage

xamarin-storage has two important details to take into account:

1 not all labs have access to it
2 it will disappear in the future

Until xamarin-storage is fully removed, we allow certain pipelines to use it. This is done via the 

> 'useXamarinStorage' parameter in the templates. To make things easier to mantain, this value is set

Reviewer 2: mantain -> maintain

in the variable templates per lab. Unfortunatly, variables do not take types (boolean in this case)
so we set a string to be 'true' or 'false' and we use eq to convert it to a boolean and pass it to
the templates:

```yml
stages:
- template: templates/device-tests-stage.yml
  parameters:
    iOSDevicePool: ${{ variables.iOSDevicePool }}
    WindowsDevicePool: ${{ variables.WindowsDevicePool }}
    useXamarinStorage: eq('${{ variables.useXamarinStorage }}', 'true')
    testsLabels: '--label=run-tvos-tests,run-non-monotouch-tests,run-monotouch-tests,run-mscorlib-tests'
    statusContext: 'VSTS: device tests tvOS (DDFun)'
    iOSDeviceDemand: 'tvos' 
```

In the above example we are interested in:

```yml
  useXamarinStorage: eq('${{ variables.useXamarinStorage }}', 'true')
```

> eq will return true of false in the string comparison and that way we do have the correct type.

Reviewer 2: of -> or
